ccx_messaging/engines/s3_upload_engine.py

```diff
@@ -101,9 +100,30 @@ def process(self, broker, local_path):
         for w in self.watchers:
             w.watch_broker(broker)
 
-        if broker["cluster_id"] is None:
-            del broker["cluster_id"]
-            broker["cluster_id"] = extract_cluster_id(local_path)
+        # Always fire on_extract for archive type detection (needed for metrics)
+        # If cluster_id is None, also extract it in the same pass to avoid opening the archive twice
+        try:
+            with tarfile.open(local_path) as tf:
+                self.fire("on_extract", None, broker, tf)
+
+                # Extract cluster_id if needed (reuse already opened archive)
+                if broker["cluster_id"] is None:
+                    del broker["cluster_id"]
+                    ID_PATH = os.path.join("config", "id")
+                    try:
+                        with tf.extractfile(ID_PATH) as id_file:
+                            broker["cluster_id"] = id_file.read().decode().strip()
+                    except KeyError as ex:
+                        raise CCXMessagingError(
+                            f"The tar in {local_path} doesn't contain cluster id"
+                        ) from ex
+        except tarfile.ReadError:
+            # Not a tarfile (e.g., JSON report) - skip gracefully
+            LOG.debug("File %s is not a tarfile, skipping on_extract", local_path)
+
+            # If cluster_id is still None and it's not a tarfile, we can't extract it
+            if broker["cluster_id"] is None:
+                raise CCXMessagingError(f"The file in {local_path} doesn't look as a tarfile")
```

Comment:
This is a logic change compared to `extract_cluster_id` as now we only raise an error if cluster_id is None